### PR TITLE
Spin projection

### PR DESCRIPTION
**Pull request by pittlerf**
Hi Bartek, now I have a version that seems to be OK, I still have to adjust the example input files and make checks.

### CODE REVIEW

File: read_input_parser_cvc.l [hunk 1]
```diff
@@ -693,25 +755,25 @@ static inline double fltlist_next_token(int * const list_end){
       }
     printf("Token=%s\n",token);
     if( strcmp(token, "Naive") == 0 ){
-      cvc::g_spinprojection_loop_type[index]=0;
+      cvc::g_LoopExtract_FilterLoopTypes[index]=0;
```
Comment (kostrzewa) [inline read_input_parser_cvc.l]: Could you use an enum for the numeric values please? 

-> ```include/enums.hpp```

```
typedef enum LoopExtractLoopType_t {
  LOOP_EXTRACT_LOOP_TYPE_NAIVE = 0,
  LOOP_EXTRACT_LOOP_TYPE_SCALAR,
  LOOP_EXTRACT_LOOP_TYPE_DOP,
  LOOP_EXTRACT_LOOP_TYPE_LOOPS,
  LOOP_EXTRACT_LOOP_TYPE_LPSDW,
  LOOP_EXTRACT_LOOP_TYPE_LOOPSCV,
  LOOP_EXTRACT_LOOP_TYPE_LPSDWCV,
  LOOP_EXTRACT_LOOP_TYPE_NLOOPTYPES
} LoopExtractLoopType_t;
```

That way it's always more or less clear what's going on. If you're very motivated you can even add comments to the above :)

File: set_default.cpp [hunk 2]
```diff
@@ -68,13 +68,28 @@ void set_default_input_values(void) {
   g_resume    = _default_resume;
   g_subtract  = _default_subtract;
   g_source_location = _default_source_location;
-  int count=snprintf(spinprojectionfname_prefix,200, "%s",  _default_spinprojectionfname_prefix);
+  int count=snprintf(g_LoopExtract_InPath,200, "%s",  _default_LoopExtract_InPath);
```
Comment (kostrzewa) [inline set_default.cpp]: I think the path and filename may need to be longer. It's easily possible to generate a filename with well over 200 characters with the deep directory structures that we have to use...

**Comment by kostrzewa** [comment]
There also seems to be a conflict which needs to be resolved.

File: read_input_parser_cvc.l [hunk 3]
```diff
@@ -642,43 +638,109 @@ static inline double fltlist_next_token(int * const list_end){
...
+  if (strcmp(yytext,"yes")==0){
+    cvc::g_LoopExtract_LegacyTraj = 1;
+    if(cvc::myverbose!=0) printf("# [read_input_parser] cvc::g_LoopExtract_LegacyTraj is on\n");
+  }
+  else if (strcmp(yytext,"no")==0){
+    cvc::g_LoopExtract_LegacyTraj = 0;
+    if(cvc::myverbose!=0) printf("# [read_input_parser] cvc::g_LoopExtract_LegacyTraj is off\n");
+  }
+}
+<LOOPEXTRACTSPINPROJECTGAMMASTRUCTURE>{CORRCONFIG} {
```
Comment (kostrzewa) [inline read_input_parser_cvc.l]: how do I specify which gamma structures?

File: example_input_files/loop.input [hunk 4]
```diff
@@ -1,32 +1,51 @@
...
+loopextractoutqsq = 4
+
+# The original loop contains momenta upto q^2=12
+loopextractinqsq = 12
+
+# We have the loops for 64 random sources
+loopextractsamples = 64
+
+# We want also ascii files to be produced
+loopextractverbositylevel = 1
```
Comment (kostrzewa) [inline example_input_files/loop.input]: can we please call this as discussed? LoopExtract_ASCII_Output

Comment (kostrzewa) [reply example_input_files/loop.input]: Also, in the example input file, can you use camel-case, just to make it easier to read...

File: read_input_parser_cvc.l [hunk 5]
```diff
@@ -693,25 +779,25 @@ static inline double fltlist_next_token(int * const list_end){
       }
     printf("Token=%s\n",token);
```
Comment (kostrzewa) [inline read_input_parser_cvc.l]: remove debugging printf

**Comment by kostrzewa** [comment]
By the way, will this thing work at all when the number of MPI tasks is larger than 1? If not, it should error out if a launch with multiple tasks is attempted.

File: read_input_parser_cvc.l [hunk 6]
```diff
@@ -422,12 +422,12 @@ static inline double fltlist_next_token(int * const list_end){
...
 ^loopextractfilenamesuffix{SPC}*={SPC}*         BEGIN(LOOPEXTRACTFILENAMESUFFIX);
-^loopextractntochaccumulated{SPC}*={SPC}*       BEGIN(LOOPEXTRACTNSTOCHACCUMULATED);
-^loopextractspinprojectspinproject{SPC}*={SPC}* BEGIN(LOOPEXTRACTSPINPROJECT);
+^loopextractnstochaccumulated{SPC}*={SPC}*      BEGIN(LOOPEXTRACTNSTOCHACCUMULATED);
+^loopextractspinproject{SPC}*={SPC}*            BEGIN(LOOPEXTRACTSPINPROJECT);
 ^loopextractlegacytraj{SPC}*={SPC}*             BEGIN(LOOPEXTRACTLEGACYTRAJ);
-^loopextractspinprojectspintrace{SPC}*={SPC}*   BEGIN(LOOPEXTRACTSPINTRACE);
+^loopextractspintrace{SPC}*={SPC}*              BEGIN(LOOPEXTRACTSPINTRACE);
 ^loopextractgammastructure{SPC}*={SPC}*         BEGIN(LOOPEXTRACTSPINPROJECTGAMMASTRUCTURE);
-^loopextractspinprojectgammabasis{EQL}          BEGIN(LOOPEXTRACTSPINPROJECTGAMMABASIS);
```
Comment (kostrzewa) [inline read_input_parser_cvc.l]: ```loopextractspinprojectgammabasis``` is perfect, the gamma basis should be completely irrelevant if no projection is done, please keep as is

File: read_input_parser_cvc.l [hunk 7]
```diff
@@ -422,12 +422,12 @@ static inline double fltlist_next_token(int * const list_end){
 ^loopextractinpath{SPC}*={SPC}*                 BEGIN(LOOPEXTRACTINPATH);
 ^loopextractfilenameprefix{SPC}*={SPC}*         BEGIN(LOOPEXTRACTFILENAMEPREFIX);
 ^loopextractfilenamesuffix{SPC}*={SPC}*         BEGIN(LOOPEXTRACTFILENAMESUFFIX);
-^loopextractntochaccumulated{SPC}*={SPC}*       BEGIN(LOOPEXTRACTNSTOCHACCUMULATED);
-^loopextractspinprojectspinproject{SPC}*={SPC}* BEGIN(LOOPEXTRACTSPINPROJECT);
+^loopextractnstochaccumulated{SPC}*={SPC}*      BEGIN(LOOPEXTRACTNSTOCHACCUMULATED);
+^loopextractspinproject{SPC}*={SPC}*            BEGIN(LOOPEXTRACTSPINPROJECT);
 ^loopextractlegacytraj{SPC}*={SPC}*             BEGIN(LOOPEXTRACTLEGACYTRAJ);
-^loopextractspinprojectspintrace{SPC}*={SPC}*   BEGIN(LOOPEXTRACTSPINTRACE);
```
Comment (kostrzewa) [inline read_input_parser_cvc.l]: LoopExtractSpinProjectSpinTrace is also good as it is

**Comment by pittlerf** [comment]
> By the way, will this thing work at all when the number of MPI tasks is larger than 1? If not, it should error out if a launch with multiple tasks is attempted.

I think, this will work with MPI as well, I can test it on the devel queue on juwels

File: read_input_parser_cvc.l (hunk 3 above)
Comment (pittlerf) [reply read_input_parser_cvc.l]: you have to set LoopExtract_GammaStructure = 0,1,2,3,4, I will test this

**Comment by kostrzewa** [comment]
The updated ```enums.hpp``` is missing.

File: include/enums.hpp [hunk 8]
```diff
@@ -13,6 +13,18 @@ typedef enum latDim_t {
 
 extern const char latDim_names[];
 
+typedef enum LoopExtractLoopType_t {
+  LOOP_EXTRACT_LOOP_TYPE_NAIVE = 1,
```
Comment (kostrzewa) [inline include/enums.hpp]: should the first one be zero or does it not matter?

Comment (kostrzewa) [reply include/enums.hpp]: If yes, then ```LOOP_EXTRACT_LOOP_TYPE_NLOOPTYPES_PLUS_ONE -> LOOP_EXTRACT_LOOP_TYPE_NLOOPTYPES```.

**Comment by kostrzewa** [comment]
Looking good, just three more things:

[ ] ```loop_extract.cpp``` -> ```filename``` should be of length 1200 or so.l
[ ] ```enums.hpp``` -> is it correct for the enum to start at 1? (this is my fault, I know...)
[ ] ```loop_extract.cpp``` -> should forcibly quit if running with more than one MPI task

Thanks!

File: loop_extract.cpp [hunk 9]
```diff
@@ -117,6 +117,13 @@ int main(int argc, char **argv) {
    * initialize MPI parameters for cvc
    ***************************************************************************/
   mpi_init(argc, argv);
+#ifdef HAVE_MPI
+  if (g_nproc > 1){
+    fprintf(stderr,"# [loop_extract] The MPI part of the code is broken, please
+    use it with a single MPI thread\n");
```
Comment (kostrzewa) [inline loop_extract.cpp]: MPI task, a thread is something else... :)

File: include/enums.hpp [hunk 10]
```diff
@@ -14,7 +14,7 @@ typedef enum latDim_t {
 extern const char latDim_names[];
 
 typedef enum LoopExtractLoopType_t {
-  LOOP_EXTRACT_LOOP_TYPE_NAIVE = 1,
+  LOOP_EXTRACT_LOOP_TYPE_NAIVE = 0,
```
Comment (kostrzewa) [inline include/enums.hpp]: the last element should thus be called NTYPES rather than NTYPES_PLUS_ONE, right?

File: read_input_parser_cvc.l (hunk 5 above)
Comment (kostrzewa) [reply read_input_parser_cvc.l]: ping

Comment (pittlerf) [reply read_input_parser_cvc.l]: I think I removed this yesterday.

**Comment by kostrzewa** [comment]
Alright, just the conflicts left then :)

**Comment by kostrzewa** [comment]
I can't push to your rep so I can't take care of the conflicts...

**Pull request merged**